Enroll Kibana API uses Service Accounts

jkakavas opened the pull request:
This commit changes the Enroll Kibana API to create and return 
a token for this service account, instead of setting and returning the
password of the kibana_system built-in user.

jkakavas committed:
Enroll Kibana API uses Service Accounts

This change introduces a Service Account for Kibana to use when
authenticating to Elasticsearch. The Service Account with
`kibana-system` service name under the `elastic` namespace,
uses the same RoleDescriptor as the existing kibana_system
built-in user and is its functional equivalent when it comes to
AuthZ.

It also changes the Enroll Kibana API to create and return a token
for this service account, instead of setting and returning the
password of the kibana_system built-in user.

jkakavas committed:
minor updates

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

jkakavas commented on client/rest-high-level/src/main/java/org/elasticsearch/client/security/KibanaEnrollmentResponse.java:
```diff
@@ -18,21 +18,21 @@
 public final class KibanaEnrollmentResponse {
 
-    private SecureString password;
+    private SecureString token;
```
@azasypkin would Kibana want to know the token name in addition to the value ? ( https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-service-token.html )

ywangd replied:
A explicit token name might be convenient for kibana, but the token value itself also has the token name embedded. The token value takes the format of `\0\1\0\1elastic/kibana-system/$tokenName:$tokenSecret` and is base64 encoded.

azasypkin replied:
> @azasypkin would Kibana want to know the token name in addition to the value ? ( https://www.elastic.co/guide/en/elasticsearch/reference/current/security-api-create-service-token.html )

Yeah, I think it'd be better to have one (to log it least).

thomheymann replied:
From what I can gather we only need the service account token to authenticate so don't think we need the name at all.

azasypkin replied:
Right, we don't need it in the code, but I thought it'd be beneficial to log it anyway, to know which Kibana triggered generation of which token, for troubleshooting purposes.

thomheymann replied:
To be honest, for consistency, it might make sense to mirror the create service token API anyways:

```
{
  "token": {
    "name": "token1",
    "value": "AAEAAWVsYXN0aWM...vZmxlZXQtc2VydmVyL3Rva2VuMTo3TFdaSDZ" 
  }
}
```

Learn once, use anywhere, etc.

jkakavas replied:
coo, I'll make the change to return the name too, thanks for weighing in folks

jkakavas commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/service/ElasticServiceAccountsTests.java:
```diff
@@ -46,6 +145,280 @@
 
 public class ElasticServiceAccountsTests extends ESTestCase {
 
+    public void testKibanaSystemPrivileges() {
```
Not certain we need to replicate ( read: copy) all of the tests here as it uses the same roledescriptor as the kibana_system and we test this in ReservedRolesStoreTests

jkakavas committed:
fix tests

jkakavas committed:
unused import

Commit:
Update service accounts overview to include kibana-system

lockewritesdocs commented:
@jkakavas, I updated the [service accounts usage](https://www.elastic.co/guide/en/elasticsearch/reference/current/service-accounts.html#service-accounts-explanation) section in https://github.com/elastic/elasticsearch/pull/76370/commits/0180b161f4367d9aac1e1fe4b14e34e1933d780e to include the `elastic/kibana-system` service account.

jkakavas committed:
fix more tests

jkakavas committed:
Merge remote-tracking branch 'origin/master' into kibana-service-account-enroll-api

jkakavas commented:
I split the introduction of the kibana-system service account to https://github.com/elastic/elasticsearch/pull/76449 so that we can  backport that to 7.15 and I will adjust this PR to simply add the relevant changes to the enrollment API. Sorry for the confusion, I think it will be clearer this way.

jkakavas committed:
Merge remote-tracking branch 'origin/master' into kibana-service-account-enroll-api

jkakavas committed:
update service name

BigPandaToo commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SecurityDocumentationIT.java:
```diff
@@ -2918,7 +2918,7 @@ public void onFailure(Exception e) {
     }
 
-    @AwaitsFix(bugUrl = "Determine behavior for keystores with multiple keys")
+    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/75097")
```
Same as above ^^

BigPandaToo reviewed (commented):
I added a couple of comments regarding tests, otherwise looks good.

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/documentation/SecurityDocumentationIT.java:
```diff
@@ -2875,7 +2875,7 @@ public void onFailure(Exception e) {
     }
 
-    @AwaitsFix(bugUrl = "Determine behavior for keystores with multiple keys")
+    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/75097")
```
True, we should have re-enabled this test as part of https://github.com/elastic/elasticsearch/pull/73807. Will take care of it on Monday

BigPandaToo replied:
Actually, `testEnrollNode` and `testEnrollKibana` [here](https://github.com/elastic/elasticsearch/blob/master/client/rest-high-level/qa/ssl-enabled/src/javaRestTest/java/org/elasticsearch/client/EnrollmentIT.java) are already doing what we need, I believe

jkakavas replied:
No, not really. EnrollIT means to test the HLRC behavior where  the *DocumentationIT means to test that our documentation snippets are accurate and work

jkakavas committed:
re-enable enrollment documentation tests

ywangd reviewed (approved):
LGTM

azasypkin commented:
Looks good to me, @thomheymann since you have a WIP PR where we interact with the Enrollment API already can you please check if this change works well for you as well?

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportKibanaEnrollmentAction.java:
```diff
@@ -98,28 +94,19 @@
+    protected static String getTokenName(){
+        final ZonedDateTime autoConfigDate = ZonedDateTime.now(ZoneOffset.UTC);
+        final String prefix = "enroll-process-token-";
+        return prefix + autoConfigDate.toInstant().getEpochSecond();
```
not really, thought second precision is good enough for unique names ,but on second thought there is nothing stopping two kibana instances from trying to enroll at the same time. Will change to millis

thomheymann commented:
> Looks good to me, @thomheymann since you have a WIP PR where we interact with the Enrollment API already can you please check if this change works well for you as well?

Thanks for ping. Yeh, happy to get this breaking change in before release. Shouldn't be too much work to adapt the existing PR. Very exciting!

jkakavas committed:
Address feedback

- nits
- removed unecessary parsing from KibanaEnrollmentResponse
- added token name in the response
- update docs

jkakavas commented:
@ywangd thanks for the comments :pray: , I think I addressed all the feedback :+1: 
@thomheymann I added the token name and updated the response as suggested, thanks, please take another look !

jkakavas committed:
update var name

jkakavas commented on client/rest-high-level/src/test/java/org/elasticsearch/client/security/KibanaEnrollmentResponseTests.java:
```diff
@@ -0,0 +1,91 @@
+
+import static org.hamcrest.Matchers.equalTo;
+
+public class KibanaEnrollmentResponseTests extends ESTestCase {
```
github fails to show this clearly but this was renamed from `KibanaErnollmentResponseTests` to fix the typo, hence the new file

jkakavas committed:
Merge remote-tracking branch 'origin/master' into kibana-service-account-enroll-api

BigPandaToo reviewed (approved):
LGTM

Commit:
Fix links, streamline grammar, and other editorial changes

lockewritesdocs commented:
> @lockewritesdocs can you please do a quick round of the doc changes ?

@jkakavas -- done in 4dce3181912d2b709dbda488623281cbe32074b0 👍

lockewritesdocs reviewed (approved):
👍 with the changes in 4dce3181912d2b709dbda488623281cbe32074b0

jkakavas commented:
Adam, it’s actually the value of the token (so <2> ) that is used to
Authenticate as the service account. <1> is just the name of the token
(returned for informational purposes). Can you please add a suggestion
based on that ?

On Monday, August 16, 2021, Adam Locke ***@***.***> wrote:

> ***@***.**** approved this pull request.
>
> 👍 with the changes in 4dce318
> <https://github.com/elastic/elasticsearch/commit/4dce3181912d2b709dbda488623281cbe32074b0>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/elasticsearch/pull/76370#pullrequestreview-731118719>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACOOCKF2ZW53IQACFXWD7UDT5F2GRANCNFSM5B642DIQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&utm_campaign=notification-email>
> .
>

Commit:
Clarify the value of the token used to authenticate

jkakavas commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into kibana-service-account-enroll-api

jkakavas commented:
@elasticmachine test this please

thomheymann reviewed (approved):
Looks great, thanks!
